### ISSUE TITLE
Autodiffbackend

### ISSUE THREAD

**Issue by wangjiawen2013** [tracel-ai/burn#2469]
Hi,
What's the difference between `NdArray` and `Autodiff<NdArray` ? I though the backend would be determined automatically when setting device as `NdArrayDevice::Cpu`, but it seems that we also need to set `NdArray` or `Autodiff<NdArray>` manually.

**Comment by loganbnielsen** [comment]
Autodiff implements Backend and so does NdArray. But Autodiff backend has an inner backend that knows how to do the computation. (Autodiff is just for computing gradients). A device is not a backend but is associated with a backend

```rust
pub trait Backend:  // <-- A backend
    FloatTensorOps<Self>
    + BoolTensorOps<Self>
    + IntTensorOps<Self>
    ...
    + 'static
{
    /// Device type.
    type Device: DeviceOps; // <-- has an associated device
```
```rust
        type Backend = Autodiff<Wgpu>;  // <-- specify backend
        let device = WgpuDevice::default(); // <-- specify device 

        let x: Tensor<Backend, 1> = Tensor::from_floats([3.0], &device);
        let mut y: Tensor<Backend, 1> = Tensor::from_floats([2.0], &device);
```

**Comment by wangjiawen2013** [comment]
As "A device is not a backend but is associated with a backend", can burn infer the backend from the device ? 
In the example above, the device is `WgpuDevice::default();` , is the compiler smart enough to infer that the backend is `wgpu` because `WgpuDevice` is used ?

**Comment by wangjiawen2013** [comment]
I tested the following code:
```rust
let device = NdArrayDevice::default();
let x: Tensor<NdArray, 1> = Tensor::from_floats([3.0], &device);
let x: Tensor<Autodiff<NdArray>, 1> = Tensor::from_floats([3.0], &device);
```
Both `NdArray` and `Autodiff<NdArray>` works. So the compiler cannot determine the backend is `NdArray` or `Autodiff<NdArray>`, we must set it explicitly, even the `NdArrayDevice::default()` is given. Am I right ?

**Comment by laggui** [comment]
Yep! The device can be inferred from the backend, but not the other way around. `NdArray` is a concrete backend implementation, and `Autodiff` also implements `Backend` but uses the inner backend to perform the tensor operations and adds auto-differentiation capabilities to the backend used. Autodiff is intended to be used as a decorator/wrapper, which means you have to be explicit about when to use it. This is intended so you specify explicitly when you want to track the operations graph, otherwise you'd be wasting resources for nothing 🙂 

See also: https://github.com/tracel-ai/burn/discussions/2415
